### PR TITLE
Implement augmentation routines in metatomic

### PR DESCRIPTION
**Pull request by ppegolo**
Augmentation is currently implemented in metatrain, and it depends on `spherical` for augmentation of spherical targets. 
In prepraration for the `SymmetrizedModel` PR (currently #119, that would depend on metatrain just for the augmentation routines) I re-implemented the relevant parts of [spherical](https://github.com/moble/spherical) (MIT license) here, getting rid of unnecessary stuff and most dependencies, and all the necessary functions that handle augmentation.

# Contributor (creator of pull-request) checklist

 - [x] Tests updated (for new features and bugfixes)?
 - [x] Documentation updated (for new features)?
 - [ ] Issue referenced (for PRs that solve an issue)?

# Reviewer checklist

 - [ ] CHANGELOG updated with public API or any other important changes?

[📚 Download documentation for this pull-request](https://nightly.link/metatensor/metatomic/actions/artifacts/8166158224.zip)

[⚙️ Download Python wheels for this pull-request (you can install these with pip)](https://nightly.link/metatensor/metatomic/actions/artifacts/8166304299.zip)

### CODE REVIEW

File: python/metatomic_torch/metatomic/torch/_augmentation/_wigner.py [hunk 1]
```diff
@@ -0,0 +1,527 @@
+"""Private Wigner-d/Wigner-D helpers for symmetry operations.
```
Comment (Luthaf) [inline python/metatomic_torch/metatomic/torch/_augmentation/_wigner.py]: I wondering if this would be better inside https://github.com/Luthaf/wigners? but OK for now

File: python/metatomic_torch/metatomic/torch/o3/_wigner.py
Comment (Luthaf) [inline python/metatomic_torch/metatomic/torch/o3/_wigner.py:1]: maybe move this to `metateomic/torch/_augmentation/_wigner.py`

File: python/metatomic_torch/tests/augmentation.py
Comment (Luthaf) [inline python/metatomic_torch/tests/augmentation.py:1]: this is a bit hard to follow, would be nice to make it clearer/better documented

File: python/metatomic_torch/metatomic/torch/_augmentation.py [hunk 2]
```diff
@@ -0,0 +1,545 @@
...
+import metatensor.torch as mts
+import numpy as np
+import torch
+from metatensor.torch import TensorBlock, TensorMap
+
+from . import System, register_autograd_neighbors
+from ._wigner import compute_wigner_batch
+
+
+def _block_row_indices_by_system(
```
Comment (Luthaf) [inline python/metatomic_torch/metatomic/torch/_augmentation.py]: Is this faster than doing `samples.select(Labels("system", [[i]]))`?

File: python/metatomic_torch/metatomic/torch/_augmentation.py [hunk 3]
```diff
@@ -0,0 +1,545 @@
...
+        types=system.types,
+        cell=system.cell @ transformation.T,
+        pbc=system.pbc,
+    )
+    for data_name in system.known_data():
+        data = system.get_data(data_name)
+        if len(data) != 1:
+            raise ValueError(
+                f"System data '{data_name}' has {len(data)} blocks; "
+                "only single-block data is supported."
```
Comment (Luthaf) [inline python/metatomic_torch/metatomic/torch/_augmentation.py]: why?

File: python/metatomic_torch/metatomic/torch/_augmentation.py [hunk 4]
```diff
@@ -0,0 +1,545 @@
...
+                            values=(
+                                block.values.swapaxes(-1, -2) @ transformation.T
+                            ).swapaxes(-1, -2),
+                            samples=block.samples,
+                            components=block.components,
+                            properties=block.properties,
+                        )
+                    ],
+                ),
+            )
```
Comment (Luthaf) [inline python/metatomic_torch/metatomic/torch/_augmentation.py]: This should use `_transform_tmap`

File: python/metatomic_torch/metatomic/torch/_augmentation.py [hunk 5]
```diff
@@ -0,0 +1,545 @@
...
+    return list(R.unbind(0))
+
+
+def apply_augmentations(
+    systems: list[System],
+    targets: dict[str, TensorMap],
+    transformations: list[torch.Tensor],
+    extra_data: dict[str, TensorMap] | None = None,
+) -> tuple[list[System], dict[str, TensorMap], dict[str, TensorMap]]:
+    """Apply a batch of O(3) transformations to systems and TensorMaps simultaneously.
```
Comment (Luthaf) [inline python/metatomic_torch/metatomic/torch/_augmentation.py]: the transformations are rotation + inversion, right? I'm wondering if a better API would be expressed in term of transformations or rotations, instead of calling these "augmentations".

What you can do in the end is the same, just makes it a bit clearer when looking for functionality

**Review by Luthaf — COMMENTED** [review]
Looks good overall!

File: python/metatomic_torch/metatomic/torch/_augmentation/_wigner.py (hunk 1 above)
Comment (ppegolo) [reply python/metatomic_torch/metatomic/torch/_augmentation/_wigner.py]: Maybe for later, yes, unless you think it's trivial?

File: python/metatomic_torch/metatomic/torch/_augmentation.py (hunk 3 above)
Comment (ppegolo) [reply python/metatomic_torch/metatomic/torch/_augmentation.py]: No reason, now it's fine

File: python/metatomic_torch/metatomic/torch/_augmentation/_wigner.py (hunk 1 above)
Comment (Luthaf) [reply python/metatomic_torch/metatomic/torch/_augmentation/_wigner.py]: Trivial is a strong word, but I might give it a quick try. Otherwise we can merge this and change it later

Comment (Luthaf) [reply python/metatomic_torch/metatomic/torch/_augmentation/_wigner.py]: It was not that bad, I think I can have a release by the end of the day https://github.com/Luthaf/wigners/pull/18

Comment (Luthaf) [reply python/metatomic_torch/metatomic/torch/_augmentation/_wigner.py]: Can you check if the Python API makes sense for this use-case? https://github.com/Luthaf/wigners/pull/18/files#diff-c87cace15e30e245e07e40b204426e6d8246a567838786e0fb924f8ffc99fe93R96

Comment (ppegolo) [reply python/metatomic_torch/metatomic/torch/_augmentation/_wigner.py]: Thank you! At a first glance I think it could substitute `_compute_wigner_d_complex`, so in the current state it should still be massaged to be used here, but it's probably fine. Not sure I have time to test it today, but I'll try.

Comment (ppegolo) [reply python/metatomic_torch/metatomic/torch/_augmentation/_wigner.py]: Tested, it works!

File: python/metatomic_torch/metatomic/torch/_augmentation.py (hunk 2 above)
Comment (ppegolo) [reply python/metatomic_torch/metatomic/torch/_augmentation.py]: Yes, a bit more than twice as fast

File: python/metatomic_torch/tests/augmentation.py
Comment (ppegolo) [reply python/metatomic_torch/tests/augmentation.py:1]: hopefully clearer now

**Comment by Luthaf** [comment]
I think I'm happy with the new API, I'll let you check if this is fine for you as well @ppegolo. Main changes are 

- no augmentation/dataset handling, only `transform_{system,tensor,block}`
- new O3Transformation class to group the transformation matrix & wigner_D matrix cache
- the code no longer try to guess the system indices. They should either match `list(range(len(systems)))`, or be passed explicitly

Still TBD before merging:

- [x] add docstring for the different modules
- [x] Add some explanation in the docs of the convention we follow for everything (real sph, naming using `o3_`)

**Review by Luthaf — APPROVED** [review]
